src/game/client/draw/w3dmodeldraw.cpp

```diff
@@ -1154,3 +1179,1913 @@ void W3DModelDrawModuleData::Parse_Condition_State(INI *ini, void *instance, voi
+                    launch_pos->Pre_Rotate_Z(info->m_turretInfo[tur].m_turretArtAngle);
+                    launch_pos->Pre_Rotate_Y(-info->m_turretInfo[tur].m_turretArtPitch);
+                }
+            }
+        }
+
+        if (turret_rot_pos) {
+            turret_rot_pos->Zero();
+        }
+
+        if (turret_pitch_pos) {
+            turret_pitch_pos->Zero();
+        }
+
+        if (tur != TURRET_INVALID) {
+            Vector3 *offset = data->Get_Attach_To_Drawable_Bone_Offset(Get_Drawable());
+            ModelConditionInfo::TurretInfo &tinfo = info->m_turretInfo[tur];
+
+            if (turret_rot_pos) {
+                bool b = tinfo.m_turretAngleName && !info->Find_Pristine_Bone_Pos(tinfo.m_turretAngleName, *turret_rot_pos);
+                captainslog_dbgassert(!b,
+                    "*** ASSET ERROR: TurretBone %s not found!",
+                    g_theNameKeyGenerator->Key_To_Name(tinfo.m_turretAngleName).Str());
+
+                if (offset) {
+                    turret_rot_pos->x += offset->X;
+                    turret_rot_pos->y += offset->Y;
+                    turret_rot_pos->z += offset->Z;
+                }
+            }
+
+            if (turret_pitch_pos) {
+                bool b = tinfo.m_turretPitchName && !info->Find_Pristine_Bone_Pos(tinfo.m_turretPitchName, *turret_rot_pos);
+                captainslog_dbgassert(!b,
+                    "*** ASSET ERROR: TurretBone %s not found!",
+                    g_theNameKeyGenerator->Key_To_Name(tinfo.m_turretPitchName).Str());
+
+                if (offset) {
+                    turret_pitch_pos->x += offset->X;
+                    turret_pitch_pos->y += offset->Y;
+                    turret_pitch_pos->z += offset->Z;
+                }
+            }
+        }
+
+        return launch_pos != nullptr;
+    } else {
+        return false;
+    }
+}
+
+int W3DModelDraw::Get_Pristine_Bone_Positions_For_Condition_State(BitFlags<MODELCONDITION_COUNT> const &c,
+    char const *bone_name,
+    int start_index,
+    Coord3D *positions,
+    Matrix3D *transforms,
+    int max_bones) const
+{
+    const ModelConditionInfo *info = Find_Best_Info(c);
+
+    if (info == nullptr) {
+        return 0;
+    }
+
+    RenderObjClass *robj;
+
+    if (info == m_curState) {
+        robj = m_renderObject;
+    } else {
+        robj = nullptr;
+    }
+
+    info->Validate_Stuff(robj, Get_Drawable()->Get_Scale(), Get_W3D_Model_Draw_Module_Data()->m_extraPublicBone);
+    Matrix3D t[64];
+
+    if (max_bones > 64) {
+        max_bones = 64;
+    }
+
+    if (!transforms) {
+        transforms = t;
+    }
+
+    int current = 0;
+    int count = start_index != 0 ? 99 : 0;
+
+    for (int i = start_index; i <= count; i++) {
+        char buf[256];
+
+        if (i) {
+            sprintf(buf, "%s%02d", bone_name, i);
+        } else {
+            strcpy(buf, bone_name);
+        }
+
+        for (char *j = buf; j && *j; j++) {
+            *j = tolower(*j);

```

Comment:
Something is not right. In `Validate_Weapon_Barrel_Info` function it generates NameKey without lower. And here it lowers. This looks like it could cause bug somewhere.

Reply:
Nope, both functions are correct, I checked.

Reply:
yeah but add a todo comment to investigate, its indeed very weird that its inconsistent here

Reply:
Its not weird because the values used by Validate_Weapon_Barrel_Info are made lower case by Parse_Weapon_Bone_Name when they are read from the ini.